File: src/main/java/com/yanli/flink/java/streamingApi/mysql/FlinkConnectMysql.java
package com.yanli.flink.java.streamingApi.mysql;

import org.apache.flink.api.java.io.jdbc.JDBCInputFormat;
import org.apache.flink.api.java.io.jdbc.JDBCInputFormat.JDBCInputFormatBuilder;
import org.apache.flink.api.java.io.jdbc.JDBCOutputFormat;
import org.apache.flink.api.java.io.jdbc.JDBCOutputFormat.JDBCOutputFormatBuilder;
import org.apache.flink.api.java.typeutils.RowTypeInfo;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.types.Row;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.sql.Types;
import java.util.List;

/**
 * @author YanLi
 * @version 1.0
 * @ClassName: FlinkConnectMysql
 * @date 2021/2/24 5:12 下午
 */
public class FlinkConnectMysql {
    private static final Logger logger = LoggerFactory.getLogger(FlinkConnectMysql.class);

    /**
     *  查询mysql 数据
     * @param env
     * @param tableName
     * @param rowTypeInfo
     * @return
     */
    public static DataStream<Row> selectAll(StreamExecutionEnvironment env, String tableName, RowTypeInfo rowTypeInfo){
        JDBCInputFormatBuilder inputFormatBuilder = FlinkJDBCFormat.getJDBCInputFormat();
        JDBCInputFormat inputFormat = inputFormatBuilder.setQuery("SELECT * FROM " + tableName)
                .setRowTypeInfo(rowTypeInfo)
                .finish();
        DataStream<Row> dataStream = env.createInput(inputFormat);

        return dataStream;
    }

    /**
     * 功能描述:
     * 〈向 表里插入数据〉
     * @Param: [tableName, rows]
     * @Return: void
     * @Author: yanli
     * @Date: 2021/2/22 11:32 上午
     */
    public static void insertTable(String tableName, List<Row> rowList){
        JDBCOutputFormatBuilder outputFormatBuilder = FlinkJDBCFormat.getJDBCOutputFormat();
        JDBCOutputFormat outputFormat = outputFormatBuilder.setQuery("insert into " + tableName + "values(?,?,?)")
                .setSqlTypes(new int[]{Types.INTEGER, Types.VARCHAR, Types.DOUBLE})
                .finish();
        try {
            outputFormat.open(1,1);
            for(Row row:rowList){
                outputFormat.writeRecord(row);
            }
            outputFormat.close();
        } catch (IOException e) {
            e.printStackTrace();
            logger.error("Close HBase Exception:", e.toString());
        }

    }

    /**
     * 功能描述:
     * 〈更新表中的一条数据〉
     * @Param: [tableName, row]
     * @Return: void
     * @Author: yanli
     * @Date: 2021/2/22 11:32 上午
     */
    public static void updataTable(String tableName, Row row){
        JDBCOutputFormatBuilder outputFormatBuilder = FlinkJDBCFormat.getJDBCOutputFormat();
        JDBCOutputFormat outputFormat = outputFormatBuilder.setQuery("update " + tableName + " set name = ?, password = ? where id = ?")
                .setSqlTypes(new int[]{Types.INTEGER, Types.VARCHAR, Types.DOUBLE})
                .finish();
        try {
            outputFormat.open(1,1);

            outputFormat.writeRecord(row);

            outputFormat.close();
        } catch (IOException e) {
            e.printStackTrace();
            logger.error("Close HBase Exception:", e.toString());
        }
    }
}

File: src/main/scala/com/yanli/flink/scala/streaming/mysql/FlinkConnectMysql.scala
package com.yanli.flink.scala.streaming.mysql

import java.sql.Types

import org.apache.flink.api.java.io.jdbc.{JDBCInputFormat, JDBCOutputFormat}
import org.apache.flink.api.java.typeutils.RowTypeInfo
import org.apache.flink.streaming.api.scala.{DataStream, StreamExecutionEnvironment}
import org.apache.flink.api.scala._
import org.apache.flink.types.Row


/**
 * @version 1.0
 * @ClassName: FlinkConnectMysql 
 * @author YanLi
 * @date 2020/1/20 4:29 下午
 */
object FlinkConnectMysql {
  /*
   * 功能描述:
   * 〈查询mysql 表〉
   * @Param: [env, tableName, rowTypeInfo]
   * @Return: org.apache.flink.streaming.api.scala.DataStream<org.apache.flink.types.Row>
   * @Author: yanli
   * @Date: 2020/1/20 4:37 下午
   */
  def selectAll(env:StreamExecutionEnvironment,tableName: String,rowTypeInfo: RowTypeInfo): DataStream[Row] ={
    val inputFormatBuilder = FlinkJDBCFormat.getJDBCInputFormat()
    val inputFormat: JDBCInputFormat = inputFormatBuilder.setQuery("SELECT * FROM " + tableName)
      .setRowTypeInfo(rowTypeInfo)
      .finish()
    val rowData: DataStream[Row] = env.createInput(inputFormat)
    rowData
  }

  /*
   * 功能描述:
   * 〈向 表里插入数据〉
   * @Param: [tableName, rows]
   * @Return: void
   * @Author: yanli
   * @Date: 2020/1/20 4:47 下午
   */
  def insertTable(tableName: String,rows:Array[Row]): Unit ={
    val outputFormatBuilder: JDBCOutputFormat.JDBCOutputFormatBuilder = FlinkJDBCFormat.getJDBCOutputFormat()
    val outputFormat: JDBCOutputFormat = outputFormatBuilder.setQuery("insert into " + tableName + "values(?,?,?)")
      .setSqlTypes(Array[Int](Types.INTEGER, Types.VARCHAR, Types.DOUBLE))
      .finish()

    outputFormat.open(1,1)

    for (row <- rows){
      outputFormat.writeRecord(row)
    }

    outputFormat.close()
  }

  /*
   * 功能描述:
   * 〈更新表中的一条数据〉
   * @Param: [tableName, row]
   * @Return: void
   * @Author: yanli
   * @Date: 2020/1/20 4:51 下午
   */
  def updataTable(tableName: String,row: Row): Unit ={
    val outputFormatBuilder: JDBCOutputFormat.JDBCOutputFormatBuilder = FlinkJDBCFormat.getJDBCOutputFormat()
    val outputFormat: JDBCOutputFormat = outputFormatBuilder.setQuery("update "+tableName+" set name = ?, password = ? where id = ?")
      .setSqlTypes(Array[Int](Types.INTEGER, Types.VARCHAR, Types.DOUBLE))
      .finish()
    outputFormat.open(1,1)

    outputFormat.writeRecord(row)

    outputFormat.close()
  }

}
